### PR TITLE
refactor: deprecated slot syntax [2117]

### PR DESCRIPTION
**Pull request by Tirokk**
> [rbeer](https://github.com/rbeer) **Authored by [rbeer](https://github.com/rbeer)**
_<time datetime="2020-03-11T17:08:06Z" title="Wednesday, March 11th 2020, 6:08:06 pm +01:00">Mar 11, 2020</time>_
_Merged <time datetime="2020-03-16T09:30:10Z" title="Monday, March 16th 2020, 10:30:10 am +01:00">Mar 16, 2020</time>_
---

## 🍰 Pullrequest

> Following https://vuejs.org/v2/guide/components-slots.htmlNamed-Slots
We are using deprecated syntax in various places in our code base
In short, <template slot="index" slot-scope="scope"\> can be refactored to <template #index="scope"\>
Motivation for this change, and more examples can be found here https://github.com/vuejs/rfcs/blob/master/active-rfcs/0001-new-slot-syntax.md

### Issues

- addresses #2117 

### Todo
1. [x] use shorthand for default slots
2. [x] ~Some test failures are seemingly caused by <ds-*> not handling the new syntax well. If so, and since they are supposed to be refactored out anyway, this might be a blocker.~
Turns out, components are not the cause. Just didn't understand the new syntax, fully. ([see comment](https://github.com/Human-Connection/Human-Connection/pull/3258#issuecomment-598454126))

### CODE REVIEW

**Comment by Tirokk** [comment]
> [cypress[bot]](https://github.com/apps/cypress) **Authored by [cypress[bot]](https://github.com/apps/cypress)**
_<time datetime="2020-03-11T17:32:41Z" title="Wednesday, March 11th 2020, 6:32:41 pm +01:00">Mar 11, 2020</time>_

---



<p align="center">
  <br />
  
</p>

---

### Test summary
<a href="https://dashboard.cypress.io/projects/qa7fe2/runs/6083/specs"></a> <b title="Passed">66</b>  &bull; <a href="https://dashboard.cypress.io/projects/qa7fe2/runs/6083/failures"></a> <b title="Failed">0</b>  &bull;  <b title="Pending">0</b> &bull;  <b title="Skipped">0</b>

---

### Run details
| | |
| --- | --- |
| **Project** | Human-Connection |
| **Status** | Passed |
| **Commit** | c2e8ea0634 |
| **Started** | Mar 14, 2020 12:23 AM |
| **Ended** | Mar 14, 2020 12:44 AM |
| **Duration** | 21:31 <a href="https://on.cypress.io/parallelization" title="Enable parallelization to reduce run duration">💡</a> |
| **OS** | Linux Ubuntu Linux - 16.04 |
| **Browser** | Firefox 68  |

[View run in Cypress Dashboard ➡️](https://dashboard.cypress.io/projects/qa7fe2/runs/6083/specs)



---

This comment has been generated by **cypress-bot** as a result of this project's GitHub integration settings. You can manage this integration in [this project's settings in the Cypress Dashboard](https://dashboard.cypress.io/projects/qa7fe2/settings)

**Comment by Tirokk** [comment]
> [cypress[bot]](https://github.com/apps/cypress) **Authored by [cypress[bot]](https://github.com/apps/cypress)**
_<time datetime="2020-03-11T17:52:25Z" title="Wednesday, March 11th 2020, 6:52:25 pm +01:00">Mar 11, 2020</time>_

---



<p align="center">
  <br />
  
</p>

---

### Test summary
<a href="https://dashboard.cypress.io/projects/qa7fe2/runs/6082/specs"></a> <b title="Passed">66</b>  &bull; <a href="https://dashboard.cypress.io/projects/qa7fe2/runs/6082/failures"></a> <b title="Failed">0</b>  &bull;  <b title="Pending">0</b> &bull;  <b title="Skipped">0</b>

---

### Run details
| | |
| --- | --- |
| **Project** | Human-Connection |
| **Status** | Passed |
| **Commit** | 496c8b7adf <span style="cursor:help;" title="Your CI provider generated this commit by merging this branch into the target branch.">ℹ️</span> |
| **Started** | Mar 14, 2020 12:22 AM |
| **Ended** | Mar 14, 2020 12:42 AM |
| **Duration** | 20:15 <a href="https://on.cypress.io/parallelization" title="Enable parallelization to reduce run duration">💡</a> |
| **OS** | Linux Ubuntu Linux - 16.04 |
| **Browser** | Firefox 68  |

[View run in Cypress Dashboard ➡️](https://dashboard.cypress.io/projects/qa7fe2/runs/6082/specs)



---

This comment has been generated by **cypress-bot** as a result of this project's GitHub integration settings. You can manage this integration in [this project's settings in the Cypress Dashboard](https://dashboard.cypress.io/projects/qa7fe2/settings)

**Comment by Tirokk** [comment]
> [rbeer](https://github.com/rbeer) **Authored by [rbeer](https://github.com/rbeer)**
_<time datetime="2020-03-12T22:13:02Z" title="Thursday, March 12th 2020, 11:13:02 pm +01:00">Mar 12, 2020</time>_

---


Looks like the new Syntax is truly breaking.

tl;dr/ The new syntax requires everything to be wrapped in <template\>s when using slot scopes, because they changed `slotProps` being available *to* the component they are defined on to only being available *in* them.

From the Vue.js docs, [Compilation Scope](https://vuejs.org/v2/guide/components-slots.html#Compilation-Scope):
>As a rule, remember that:
>Everything in the parent template is compiled in parent scope; everything in the child template is compiled in the child scope.

Further down, in the section [Named Scope](https://vuejs.org/v2/guide/components-slots.html#Named-Slots) it moreover says:
>Note that v-slot can only be added to a <template\> (with one exception), unlike the deprecated slot attribute.

So I was actually wrong, @mattwr18, when I said that `v-slot` wouldn't work only on _native_ tags and usage on components being fine.

---

Changing
```js
<ds-menu :routes="routes">
    <ds-menu-item
      #menuitem="item" // defining `item` here makes it available -in- <ds-menu-item>
      :route="item.route" // ! but not -on- itself
      :parents="item.parents"
      @click.stop.prevent="openItem(item.route, toggleMenu)"
    >
      <base-icon :name="item.route.icon" />
      {{ item.route.label }}
    </ds-menu-item>
</ds-menu>
```
to
```js
<ds-menu :routes="routes">
  <template #menuitem="item"> // define `item` in a parent scope
    <ds-menu-item
      :route="item.route" // and <ds-menu-tem> has access again
      :parents="item.parents"
      @click.stop.prevent="openItem(item.route, toggleMenu)"
    >
      <base-icon :name="item.route.icon" />
      {{ item.route.label }}
    </ds-menu-item>
  </template>
</ds-menu>
```

makes the tests for [components/ContentMenu/ContentMenu.vue](https://github.com/Human-Connection/Human-Connection/blob/master/webapp/components/ContentMenu/ContentMenu.vue) pass; explaining why `item` is undefined, thus throwing `TypeError: Cannot read property 'route' of undefined`.

The question remains whether this is a good change, looking at the fact that all these <ds-*> components are about to be refactored anyways.

**Comment by Tirokk** [comment]
> [rbeer](https://github.com/rbeer) **Authored by [rbeer](https://github.com/rbeer)**
_<time datetime="2020-03-12T23:57:07Z" title="Friday, March 13th 2020, 12:57:07 am +01:00">Mar 13, 2020</time>_

---


thihi :relaxed: :tada:

**Review by Mogge — CHANGES_REQUESTED** [review]
hey this is great @rbeer 
I think it's a good change. We don't know when the `ds-menu` will get migrated, when it does there will be a bigger refactor, but the `slots` should stay with the new syntax.
Only little thing I would ask, if you don't mind, is to update our docs here https://github.com/Human-Connection/Human-Connection/blob/master/webapp/vue.md#we-use-shorthands to include the new short hand we have introduced with a link to the docs.

**Review by Mogge — APPROVED** [review]
great! thanks @rbeer !!!
